common/src/main/java/com/wynntils/core/text/StyledText.java

```diff
@@ -0,0 +1,487 @@
+    }
+
+    public static StyledText fromString(String codedString) {
+        List<StyledTextPart> parts = new LinkedList<>();
+        List<ClickEvent> clickEvents = new LinkedList<>();
+        List<HoverEvent> hoverEvents = new LinkedList<>();
+
+        Style currentStyle = Style.EMPTY;
+        StringBuilder currentString = new StringBuilder();
+
+        boolean nextIsFormatting = false;
+
+        for (char current : codedString.toCharArray()) {
+            if (nextIsFormatting) {
+                nextIsFormatting = false;
+
+                ChatFormatting formatting = ChatFormatting.getByCode(current);
+
+                if (formatting == null) {
+                    currentString.append(ChatFormatting.PREFIX_CODE);
+                    currentString.append(current);
+                    continue;
+                }
+
+                // We already had some text with the current style
+                // Append it before modifying the style
+                if (!currentString.isEmpty()) {
+                    parts.add(new StyledTextPart(currentString.toString(), currentStyle, null, null));
+                    currentString = new StringBuilder();
+                }
+
+                // Color formatting resets the style
+                if (formatting.isColor()) {
+                    currentStyle = Style.EMPTY.withColor(formatting);
+                    continue;
+                }
+
+                currentStyle = currentStyle.applyFormat(formatting);
+
+                continue;
+            }
+
+            if (current == ChatFormatting.PREFIX_CODE) {
+                nextIsFormatting = true;
+                continue;
+            }
+
+            currentString.append(current);
+        }
+
+        // Check if we have some text left
+        if (!currentString.isEmpty()) {
+            parts.add(new StyledTextPart(currentString.toString(), currentStyle, null, null));
+        }
+
+        return new StyledText(parts, clickEvents, hoverEvents);
+    }
+
+    // We don't want to expose the actual string to the outside world
+    // If you need to do an operation with this string, implement it as a method
+    public String getString(PartStyle.StyleType type) {
+        if (type == PartStyle.StyleType.FULL) {
+            return ComponentUtils.getCoded(temporaryWorkaround);
+        }
+
+        StringBuilder builder = new StringBuilder();
+
+        PartStyle previousStyle = null;
+        for (StyledTextPart part : parts) {
+            builder.append(part.getString(previousStyle, type));
+            previousStyle = part.getPartStyle();
+        }
+
+        return builder.toString();
+    }
+
+    public MutableComponent getComponent() {
+        if (parts.isEmpty()) {
+            return Component.empty();
+        }
+
+        MutableComponent component = parts.get(0).getComponent();
+
+        for (int i = 1; i < parts.size(); i++) {
+            component.append(parts.get(i).getComponent());
```

Comment:
You are setting parts to a LinkedList. This means this loop will be O(n^2). 

But I think this implementation is the correct one. What you should do is to not use LinkedList. It is worse than ArrayList in 99 cases out of 100, unless you *really* need to be able to pop of the head of the list, and make sure you know what you are doing.

Reply:
Using ArrayList everywhere now.